### PR TITLE
 basic programm

### PR DESCRIPTION
**Pull request by greleuniforme**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
 # Egg
 
-The intaller of Raven-OS
+The installer of Raven-OS
+
+First you need to install Python development for the command msgfmt3
```
Comment (Arignir) [inline README.md]: This command doesn't look like it is available on other systems but Fedora/RHEL. Please use something else, or the user won't be able to build the project (which is a big no-no).

If, and only if, the license of `msgfmt3` permits you, you can redistribute it in this repository (given that the file isn't huge).

File: README.md [hunk 2]
```diff
@@ -1,3 +1,8 @@
...
+
+At the racine of the repository launch:
```
Comment (Arignir) [inline README.md]:
```suggestion
At the root of the repository, run:
```

File: config/general.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
+{
+    "os_name": "Raven-OS",
+    "url_check_ip": "http://ip-api.com/json/?fields=status,countryCode,timezone,query",
+    "launch_without_root": false,
+    "default_language_code": "fr",
+    "locales_folder": "./locales",
+    "config_page": {
+        "language_live": {
+            "id": "language_live",
+            "icon": "./img/ico/page1.png"
```
Comment (Arignir) [inline config/general.json]: Those icons aren't present in this commit. Is that intended?

File: config/main_window.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (Arignir) [inline config/main_window.json:1]: I believe this file should be renamed `welcome_window.json`? Is that correct?

File: config/main_window.json [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    "window_xml_file": "./gtk_gui/xml/welcome_window.glade",
+    "window_size_x": "700",
+    "window_size_y": "500",
+    "window_fullscreen": true,
```
Comment (Arignir) [inline config/main_window.json]: Meh, nah! Make the window a bit bigger if you need it (but do not if you don't), without full screen, but centered vertically and horizontally (if that's possible) and the wallpaper will fill the remaining space.

File: config/main_window.json [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    "window_logo_bot_path": "./img/raven_title_white.png",
+    "window_list_components": [
+        "welcome_window",
+        "welcome_window_prev_btn",
+        "welcome_window_next_btn",
+        "welcome_window_left_top_logo",
```
Comment (Arignir) [inline config/main_window.json]:
```suggestion
        "welcome_window_top_left_logo",
```

(Same for the following buttons/logo/whatever)

File: config/main_window.json [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+        "welcome_window_left_bot_logo",
+        "welcome_window_left_bot_label",
+        "welcome_window_right_header_label",
```
Comment (Arignir) [inline config/main_window.json]:
```suggestion
        "welcome_window_header_right_label",
```

Same here, it's usually Y axis then X axis when naming things (we say the top-right corner, not the right-top corner)

File: gtk_gui/custom_modal.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import gi
+from gi.repository import Gtk
+from events.type_event import TypeEvent
+from interfaces.custom_modal_interface import CustomModalInterface
+gi.require_version('Gtk', '3.0')
```
Comment (Arignir) [inline gtk_gui/custom_modal.py]: You should add a blank line before that one

File: gtk_gui/custom_modal.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+
+@CustomModalInterface.register
+class CustomModalGtk(CustomModalInterface):
+    def __init__(self):
```
Comment (Arignir) [inline gtk_gui/custom_modal.py]: Is there a reason why you override the constructor and do not call the parent's constructor?

File: gtk_gui/custom_modal.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
+import gi
```
Comment (Arignir) [inline gtk_gui/custom_modal.py]: Instead of `CustomModel`, why not naming it `Popup`? (I'd like your opinion on this one, @doom )

File: gtk_gui/custom_modal.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+from interfaces.custom_modal_interface import CustomModalInterface
+gi.require_version('Gtk', '3.0')
+
+
+@CustomModalInterface.register
+class CustomModalGtk(CustomModalInterface):
+    def __init__(self):
+        pass
+
+    def launch(self, title, msg, event_type):
```
Comment (Arignir) [inline gtk_gui/custom_modal.py]: I think it's best to move those to the constructor, store them in the object and use them in the method's body.

File: main.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env python3.7
```
Comment (Arignir) [inline main.py]: Keep the shebang to `python3` to ensure the script will still work even when python 3.8+ is out, and please only use features of python 3.6, as Raven-OS doesn't support python 3.7 right now.

File: services/language_management.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+            code = lang
+
+        self.current_language = code
+        for current_file in self.translation_obj.keys():
+            self.translation_obj[current_file] = gettext.translation(
+                domain=current_file, localedir=self.locales_folder,
+                fallback=True, languages=[code])
+            self.translation_obj[current_file].install()
+
+    def print_in_lang(self, file, message_name):
```
Comment (Arignir) [inline services/language_management.py]: I would rename that to `translate()` or `translate_msg()`, as the current name is misleading: this function doesn't print anything, it returns!

File: services/language_management.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
+import os
+import gettext
+
+
+class LanguageManagement(object):
+    def __init__(self, locales_folder=None, lang=None, file=None):
+        self.locales_folder = locales_folder
+        if lang is None:
+            if "LANGUAGE" in os.environ:
```
Comment (Arignir) [inline services/language_management.py]: What if `LANGUAGE` isn't in `os.environ`?

File: services/language_management.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+        self.detailed_languages = {"en": "en_US", "fr": "fr_FR"}
+        self.translation_obj = {}
+
+        if lang is not None and file is not None and locales_folder is not None:
+            self.translation_obj[file] = gettext.translation(
+                domain=file, localedir=self.locales_folder,
+                fallback=True, languages=[lang])
+            self.translation_obj[file].install()
+
+    def set_locales_folder(self, locales_folder):
```
Comment (Arignir) [inline services/language_management.py]: It's common in Python to not create getter/setter and use instance variable instead.

File: services/language_management.py [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+            if "LANGUAGE" in os.environ:
+                lang = os.environ['LANGUAGE']
+        else:
+            os.environ['LANGUAGE'] = lang
+        self.default_language = lang
+
+        self.current_language = lang
+        self.available_languages = {"en": "english", "fr": "french"}
+        self.detailed_languages = {"en": "en_US", "fr": "fr_FR"}
+        self.translation_obj = {}
```
Comment (Arignir) [inline services/language_management.py]: That name is odd, maye `translater`? Would that make more sense?

File: containers.py [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+# services
+from services.language_management import LanguageManagement
+
+# gtk
+from gtk_gui.custom_modal import CustomModalGtk
+from gtk_gui.welcome_window import WelcomeWindowGtk
+
+
+class Configs(containers.DeclarativeContainer):
+    config = providers.Configuration('config')
```
Comment (Arignir) [inline containers.py]: Prefixing the instance variables with the name of the class is redundant

`config` -> `general`
`config_main_window_page` -> `main_window`

It's even better because it matches the file names!

File: containers.py [hunk 18]
```diff
@@ -0,0 +1,23 @@
+from dependency_injector import providers, containers
```
Comment (Arignir) [inline containers.py]: So the current pattern, if i'm not misreading anything, is to have these three classes in this file with static instance variables initialized twice: one time here, and another time in the main. It makes it hard to follow because the second initialization code isn't here, it's in standalone functions near the main.

If you want to do a singleton, you should probably follow the recommended patterns to do so, especially:
  * Use a constructor to initialize the different objects instead of having a function in the main to do that
  * Provide functions to interact with this object and answer the most common needs
  * Ensure there can only be one (and only one) instance of the object

See [here](https://python-3-patterns-idioms-test.readthedocs.io/en/latest/Singleton.html) for examples.

File: interfaces/custom_modal_interface.py [hunk 19]
```diff
@@ -0,0 +1,7 @@
+import abc
+
+
+class CustomModalInterface(metaclass=abc.ABCMeta):
```
Comment (Arignir) [inline interfaces/custom_modal_interface.py]: I'm not sure the two abstract class in `interfaces/` have much utility. What is the idea you had in mind when creating them?

File: containers.py [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+class Configs(containers.DeclarativeContainer):
+    config = providers.Configuration('config')
+    config_main_window_page = providers.Configuration('config')
+
+
+class Locales(containers.DeclarativeContainer):
+    locale_general = providers.Singleton(LanguageManagement, None, None, None)
+
+
+class GraphicGui(containers.DeclarativeContainer):
```
Comment (Arignir) [inline containers.py]: That graphic-thingy is in a common file, which creates an anomaly as you wan't common files between different front-ends (gtk, ncurses etc.) to not depend on a specific frontend.

File: gtk_gui/welcome_window.py [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+
+        self._builder.connect_signals(
+            {
+                'onDestroy': Handler.on_destroy,
+                'onButtonExit': Handler.on_destroy
+            })
+        self._component = Components(self._builder)
+
+        self.init_window()
+        self.set_full_screen_win()
```
Comment (Arignir) [inline gtk_gui/welcome_window.py]: I think you should inline those functions instead of calling them.
There name suggests they are setters, but they aren't, and unless i'm mistaken  they are only called once here.

**Review by Arignir — CHANGES_REQUESTED** [review]
The directory structure can further be improved. I'd like to recommend you the following one:
  * `config/`: configuration files
  * `locale/`: Language files
  * `ui/`: All frontends, one per frontend
    * `gtk/`: gtk frontend
      * `img/`: Images
      * `glade/`: Glade files
  * `egg/`: Common python files for all front-ends. **Make sure that no file requires any module from gtk.**
    * Put here files like `type_event.py`, `container.py`, `language_management.py` etc.
    * `interfaces/`: All UI stuff that is frontend independant. Put your abstract class here
  * `main.py`: The main should have a compile-time switch that picks the correct front-end (not a run-time one).

The key is to identify the core concepts behind your code (ui, configuration, locales etc.) and create clear boundaries in terms of directory structure. Especially, you have a restriction where you do NOT want gtk code in common code or the ncurses version of egg would require gtk, which feels off. Clear directory boundaries is the key to ensure that your common code is clear of any gtk code.

Furthermore, you want your main to have a hard switch which picks either gtk or ncurses. To make things very efficient, you want that switch to be as early as possible in the code and avoid at all cost the "if gtk do this else if ncurses do this etc." pattern in the remaining of the code base (or it will be hell on earth the day you want to add a third frontend). To do this, you need strong abstractions:
  * The hard-logic of the installer (installing things, partitioning etc.) must be in the common code
  * The UI-specific details must be in the correct ui/ folder
  * Use a `if gtk else if ncurses` once in the main, launch the correct window, and the remaining of the installer can be done from the `ui` part.
  * See the common folder as a library that the ui-specific code uses to do its thing
  * It'll be easier to have two different "mains", one in `ui/gtk`, another in `ui/ncurses`, and have the global 
 main pick the correct one, and harder to have  abstract classes everywhere and try to keep a common main instead
  * Put the common init code before the ui-specific main is called

Because you only support gtk for now, obviously default the switch to gtk.

An extra python rule i'd like you to follow: when using string literals, please use `'` if it's a technical string, and a `"` if it's an English sentence the user can see. No big deal, but it makes the project consistent with `nbuild`.

There's a mix between the name welcome window and main window. Pick "main window", and modify all occurrences of "welcome window" accordingly.

File: README.md [hunk 22]
```diff
@@ -1,3 +1,8 @@
 # Egg
 
-The intaller of Raven-OS
+The installer of Raven-OS
+
+First you need to install gettext for the command msgfmt
```
Comment (Arignir) [inline README.md:5]:
```suggestion

## Dependencies

Install the package `gettext` to use the command `msgfmt`. Depending on your distribution, it may be called `msgfmt`, `msgfmt3` or `msgfmt3.py`.
```

File: README.md [hunk 23]
```diff
@@ -1,3 +1,8 @@
...
+
+At the root of the repository, run:
+``` find ./locales -name "*.po" | while read f; do msgfmt -o ${f%.po}.mo $f; done ```
```
Comment (Arignir) [inline README.md]:
````suggestion
```bash
find ./locales -name "*.po" | while read f; do msgfmt -o ${f%.po}.mo $f; done
``` 

You may need to adapt this command depending on the alternative name `msgfmt` might have on your system.

## Running the intaller

To run Egg, type 

```bash
python3 ./egg.py
```

Note that only the root user is able to use Egg.

File: egg/language_management.py [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+        else:
+            lang = 'en'
+        self.default_language = lang
+
+        self.current_language = lang
+        self.available_languages = {'en': 'english', 'fr': 'french'}
+        self.detailed_languages = {'en': 'en_US', 'fr': 'fr_FR'}
+        self.translater = {}
+
+        if lang is not None and file is not None and self.locales_folder is not None:
```
Comment (Arignir) [inline egg/language_management.py]:
```suggestion
        if file is not None and self.locales_folder is not None:
```

(`lang` cannot be None, because you already tested if it was and assigned it a value.)

File: main.py [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+    with open(config_files_path['general'], 'r') as datafile:
+        config = json.load(datafile)
+        Configs.general.override(config)
+
+    with open(config_files_path['main_window'], 'r') as datafile:
+        config = json.load(datafile)
+        Configs.main_window.override(config)
+
+
+def main():
```
Comment (Arignir) [inline main.py]: What if no front-end are found? Can we have a nice error message?

**Review by Arignir — CHANGES_REQUESTED** [review]
Please start translation's words and sentences with a capital letter (`Quit`, `French` etc.), and do not add translations for pages that aren't in this PR.

The initialization and creation of the singletons are still separated, @doom will come in with handy tips and ideas to help you organize that.

File: egg/language_management.py [hunk 26]
```diff
@@ -0,0 +1,66 @@
+import os
+import gettext
+
+
+class LanguageManagement(object):
+    def __init__(self, config=None, file=None):
+        lang = config['default_language_code']
```
Comment (doom) [inline egg/language_management.py:7]: Since `config` defaults to `None`, `config['default_language_code']` might not work

File: egg/app.py [hunk 27]
```diff
@@ -0,0 +1,9 @@
+class App:
+    def __init__(self):
+        pass
+
+    def display_popup(self, title, msg, popup_type):
```
Comment (doom) [inline egg/app.py]: With Python 3.5, [type hints](https://docs.python.org/3/library/typing.html) were introduced in order to allow functions to specify the expected type for each parameter. You might want to look at that feature and use it ! :stuck_out_tongue:

File: egg/language_management.py [hunk 28]
```diff
@@ -0,0 +1,66 @@
...
+
+    def get_detailed_locale(self, locale):
+        return self.detailed_languages[locale]
+
+    def get_detailed_locale_country(self, locale):
+        locale_detailed = self.detailed_languages[locale]
+        begin = locale_detailed.find('_') + 1
+        end = len(locale_detailed)
+        if begin <= 0:
+            return None
```
Comment (doom) [inline egg/language_management.py]: If this is an error case (i.e. if it means that the detail name is invalid), you should raise an exception

File: egg/language_management.py [hunk 29]
```diff
@@ -0,0 +1,66 @@
...
+        if lang is not None and file is not None and self.locales_folder is not None:
+            self.translater[file] = gettext.translation(
+                domain=file, localedir=self.locales_folder,
+                fallback=True, languages=[lang])
+            self.translater[file].install()
+
+    def get_detailed_locale(self, locale):
+        return self.detailed_languages[locale]
+
+    def get_detailed_locale_country(self, locale):
```
Comment (doom) [inline egg/language_management.py]: Type hints can also be used to annotate a function's return type, the syntax would be:
```suggestion
    def get_detailed_locale_country(self, locale) -> str:
```

File: egg/language_management.py [hunk 30]
```diff
@@ -0,0 +1,66 @@
...
+
+        self.current_language = code
+        self.translater[file] = gettext.translation(
+            domain=file, localedir=self.locales_folder,
+            fallback=True, languages=[code])
+        self.translater[file].install()
+
+    def change_language_all_files(self, lang):
+        code = self.default_language
+        if lang in self.available_languages.keys():
```
Comment (doom) [inline egg/language_management.py]: You don't have to use `.keys()` here
```suggestion
        if lang in self.available_languages:
```

File: egg/language_management.py [hunk 31]
```diff
@@ -0,0 +1,66 @@
...
+
+        self.current_language = code
+        for current_file in self.translater.keys():
+            self.translater[current_file] = gettext.translation(
+                domain=current_file, localedir=self.locales_folder,
+                fallback=True, languages=[code])
+            self.translater[current_file].install()
+
+    def translate_msg(self, file, message_name):
+        if not self.translater or not self.translater[file]:
```
Comment (doom) [inline egg/language_management.py]: Just to be sure, you don't use `not self.translater[file]` to check if `file` is contained in `self.translater` right ?
If a subscripted key cannot be found, Python dictionaries do not return `None`, but raise a `KeyError`

File: ui/gtk/main_window.py [hunk 32]
```diff
@@ -0,0 +1,73 @@
...
+
+class Handler:
+    @staticmethod
+    def on_destroy(*args):
+        Gtk.main_quit()
+
+
+class Components:
+    def __init__(self, builder):
+        self._allComponents = {}
```
Comment (doom) [inline ui/gtk/main_window.py]: Python conventions dictate that variable names must follow snake_case 
```suggestion
        self._all_components = {}
```

File: egg/language_management.py [hunk 33]
```diff
@@ -0,0 +1,66 @@
...
+        locale_detailed = self.detailed_languages[locale]
+        begin = locale_detailed.find('_') + 1
+        end = len(locale_detailed)
+        if begin <= 0:
+            return None
+        return locale_detailed[begin:end]
+
+    def change_language_file(self, lang, file):
+        code = self.default_language
+        if lang in self.available_languages.keys():
```
Comment (doom) [inline egg/language_management.py]: You don't have to use `.keys()` here
```suggestion
        if lang in self.available_languages:
```

**Comment by greleuniforme** [comment]
@doom grange talked to me to talk with you for give to me some idea / what you think i should change for the singleton in the containers.

**Comment by doom** [comment]
So, I looked at the singleton thingy, and I think the use of the `dependency_injector` library is a bit overkill.
Since you load the configuration manually in the `main` function anyway, you don't really seem to need lazy-loading of the singleton instance. Could a regular global variable (protected by a getter maybe) initialized by the `load_*` functions be enough for what you want ?

**Comment by greleuniforme** [comment]
I did without the singleton instance finally. So I removed the dependency_injector library.

**Review by Arignir — APPROVED** [review]
